### ISSUE TITLE
Eliminate code sharing across projects to be split

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#1596]
We could do this as part of each issue in the Epic, but it will be simpler if we eliminate or reduce sharing first. Steps to take:

1. Identify which projects actually use source files in Common.
2. Move any files only used by one project to that project.
3. Copy files as needed to the projects that use them.

Note... projects not being split can obviously continue to use common, e.g.: framework and nunitlite.

**Comment by CharliePoole** [comment]
This is an inventory of the files in common, organized in logical groups that can be dealt with in the same way. Where a test file exists in common the file is marked with **(test)**.

I'll use these groupings in further comments to design a plan for eliminating common.

**Internal Trace**
These files are used to create internal trace logs. All of them are in the logging subdirectory. Shared by **framework** and **engine**.
 * ILogger.cs
 * InternalTrace.cs
 * InternalTraceLevel.cs
 * InternalTraceWriter.cs
 * Logger.cs

**Color Console Writer**
These files implement color output to the console and also provide interfaces to be used in creating the same output in non-console environments. Shared by **console** and **nunitlite**.
 * ColorConsole.cs
 * ColorConsoleWriter.cs
 * ColorStyle.cs
 * ExtendedTextWrapper.cs **(test)**
 * ExtendedTextWriter.cs

**Command Options**
These files implement command-line processing. Shared by **console** and **nunitlite**.
 * CommandLineOptions.cs
 * DefaultOptionsProvider.cs **(test)**
 * IDefaultOptionsProvider.cs
 * Options.cs
 * OutputSpecification.cs **(test)**
 * TestNameParser.cs **(test)**
 * TestSelectionParser.cs **(test)**
 * TestSelectionParserException.cs
 * Tokenizer.cs **(test)**

**Miscellaneous**
These files need to be dealt with individually.
 * AssemblyHelper.cs **(test)**<br/>
   Allows finding the path or directory of an assembly. Shared by **engine** and **framework**.
 * Env.cs<br/>
   Defines a small number of cross-platform values. Shared by **console** and **framework**.
 * Guard.cs<br/>
   Implements guard clauses for methods. Shared by **engine** and **framework**.
 * PackageSettings.cs<br/>
   Defines names of settings used in test packages. Shared by **console**, **engine**, **framework** and **nunitlite**.
 * XmlHelper.cs<br/>
   Direct and extension methods for working with XmlNodes. Shared by **console**, **engine**, **nunit-project-loader** and **nunit-v2-result-writer**.

**Comment by CharliePoole** [comment]
I spun off #1621 to deal with sharing between the console and nunitlite. A separate assembly seems to be the right approach in that case. Please add your comments and ideas to #1621 as well as here.

**Comment by CharliePoole** [comment]
I just marked this "easyfix". As a whole,  this issue is a pretty long slog, but many of the individual steps to take are easy fixes. Post a comment here that says what part you would like work on and how you plan to resolve it.

For example, I just fixed Env.cs. I might have written: "Except for two uses in the console runner, env.cs is only used by the framework. I plan to inline those two uses and move the file to the framework directory."

I'll update the inventory in an earlier comment with notes about what is done and what is being worked on.

**Comment by CharliePoole** [comment]
In order to allow others to work on parts of this, I think I'll post my proposals up front. Anybody can take one of them and let me know you are doing it... See following comments...

**Comment by CharliePoole** [comment]
Guard.cs is included in the console runner but not used. It is included and used by both the framework and the engine. This is a simple class that is used to check parameters at the start of a method. It's common for each application to have a class like this and that's my proposed resolution.

We can remove guard.cs from common and duplicate it in both framework and nunit.engine directories. Each of them can have its own copy. It has to be updated in each separate framework build of course. Comments? Volunteers?

**Comment by CharliePoole** [comment]
**PackageSettings.cs** is tricky. It's used by console, engine, nunitlite and framework! It contains the keys that are used for all our settings, with entries like...

```C#
public const string DefaultTimeout = "DefaultTimeout";
```

The file contains comments about which component uses each setting. On the one hand, sharing this is good, because the system won't work if the values get out of sync. On the other hand, there is a value in decoupling the components. For example, the framework could add a new setting and any runner that knows about it can use it. So long as the engine doesn't need to take any special action, it can remain completely in the dark about that setting.

Basically, these are string constants and the file simply gives us a way to easily avoid spelling errors. I think that these should be separate files in each component, trading off duplication for better decoupling. It can be removed from common and duplicated in nunit.console, nunit.engine and framework. Since nunitlite actually references the framework, we shouldn't have to add it there. Comments? Volunteers?

**Comment by CharliePoole** [comment]
**AssemblyHelper.cs**, as one might guess, has helper methods for working with assemblies. The engine only uses one of them, `GetDirectoryName`, in two locations. The file can be removed from common and duplicated in framework and engine. Then the engine copy can be substantially simplified, removing what is not used or needed. Comments? Volunteers?

**Comment by CharliePoole** [comment]
**XmlHelper** contains both normal static methods and extension methods for XmlNode. It is used in the engine, console and two extensions but the usage varies substantially in each of them. I think each of these needs it's own copy with only the members that are actually needed. Comments?

**Comment by ChrisMaddock** [comment]
@CharliePoole: I might have a look at picking off some of these. 

Did you have any thoughts on the logging classes? It looks like this has been additionally duplicated into the portable agent - it feels like it would be nice not to have 3 versions around. Seems like it's a fairly self-contained piece of functionality which could go into it's own assembly, although it would be much smaller than anything we're currently looking at.

**Comment by CharliePoole** [comment]
@ChrisMaddock Please grab any of the ones not marked as being worked on  and edit the top-level description to indicate that you are working on them.

The main problem with an assembly, other than the fact that it would be pretty small, is that we have always tried to keep the framework from having references. We made an exception for NUnit.System.Linq, but only for the .NET 2.0 build, which is probably not used so much. So I guess I would go for duplicating the classes, much as I dislike that. I worry less about the portable agent, because that's a bit experimental right now.

As I wrote in one of the other comments, some duplication is the price we are paying for flexibility and separation. So I'd say duplicate them and figure that we will come back to this at some point.

**Comment by ChrisMaddock** [comment]
Thanks Charlie, that makes sense.
